x/ecocredit/server/marketplace/buy.go

```diff
@@ -0,0 +1,241 @@
+					"order set to %t", sellOrder.DisableAutoRetire, order.DisableAutoRetire)
+			}
+
+			market, err := k.stateStore.MarketTable().Get(ctx, sellOrder.MarketId)
+			if err != nil {
+				return nil, fmt.Errorf("market id %d: %w", sellOrder.MarketId, err)
+			}
+			if order.BidPrice.Denom != market.BankDenom {
+				return nil, sdkerrors.ErrInvalidRequest.Wrapf("bid price denom does not match ask price denom: "+
+					" %s, expected %s", order.BidPrice.Denom, market.BankDenom)
+			}
+
+			// check that bid price is at least equal to ask price
+			askAmount, ok := sdk.NewIntFromString(sellOrder.AskPrice)
+			if !ok {
+				return nil, fmt.Errorf("could not convert sell order's ask price to %T: %s", sdk.Int{}, sellOrder.AskPrice)
+			}
+			if order.BidPrice.Amount.LT(askAmount) {
+				return nil, sdkerrors.ErrInsufficientFunds.Wrapf("bid price too low: got %s, needed at least %s",
+					order.BidPrice.Amount.String(), sellOrder.AskPrice)
+			}
+
+			if err = k.updateBalances(ctx, sellOrder, buyerAcc, orderAmt, *order.BidPrice); err != nil {
+				return nil, fmt.Errorf("error updating balances: %w", err)
+			}
+			if !sellOrder.DisableAutoRetire {
+				if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventRetire{
+					Retirer:    buyerAcc.String(),
+					BatchDenom: batch.BatchDenom,
+					Amount:     order.Quantity,
+					Location:   order.RetirementLocation,
+				}); err != nil {
+					return nil, err
+				}
+			} else {
+				if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventReceive{
+					Sender:         sdk.AccAddress(sellOrder.Seller).String(),
+					Recipient:      buyerAcc.String(),
+					BatchDenom:     batch.BatchDenom,
+					TradableAmount: order.Quantity,
+					RetiredAmount:  "",
+					BasketDenom:    "",
+				}); err != nil {
+					return nil, err
+				}
+			}
+		default:
+			return nil, sdkerrors.ErrInvalidRequest.Wrap("only direct buy orders are enabled at this time")
+		}
+	}
+	return &v1.MsgBuyResponse{}, nil
+}
+
+// updateBalances moves credits according to the order. it will:
+// - update a sell order, removing it if quantity becomes 0 as a result of this purchase.
+// - remove the purchaseQty from the seller's escrowed balance.
+// - add credits to the buyer's tradable/retired address (based on the DisableAutoRetire field).
+// - update the supply accordingly.
+// - send the coins specified in the bid to the seller.
+func (k Keeper) updateBalances(ctx context.Context, sellOrder *api.SellOrder, buyerAcc sdk.AccAddress, purchaseQty math.Dec, bidCoin sdk.Coin) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sellOrderQty, err := math.NewDecFromString(sellOrder.Quantity)
+	if err != nil {
+		return err
+	}
+	// since we only support direct buy orders at this time, we fail when the requested purchase amount is more than
+	// the available credits for sale, rather than partial filling the buy order.
+	sellOrderQty, err = math.SafeSubBalance(sellOrderQty, purchaseQty)
+	if err != nil {
+		return err
+	}
```

Comment:
We need a formatted error here. I think we should also be checking if the buy order has `DisablePartialFill` set to true as we do in the current implementation.

https://github.com/regen-network/regen-ledger/blob/ty/856-buy_orm/x/ecocredit/server/msg_server.go#L782-L785